### DISCUSSION TITLE
Date field timezone support: no way to reasonably set event times

### DISCUSSION THREAD

**Discussion by ssyberg** [payloadcms/payload#5635] [Feature Requests & Ideas]
### Describe the Bug

Scenario:
- You're in Hawaii (UTC -10)
- You are entering an event in your CMS that takes place in Moscow (UTC +3)
- The event takes place on 4/4/24 at 8am

I suggest there's effectively no reasonable way to represent this in payload. To display this on the frontend I believe you must set a timezone on this event (perhaps there's an argument that this could be business logic gleaned from an event location or that the frontend should always display the event time in the browsers local timezone but I'd say those are both pretty unreaonsable). 

The admin experience here requires some extremely counter intuitive behavior to get this correct. As an admin I must understand that whatever date I enter here will be stored in UTC, despite what timezone I select in the timezone dropdown. In order to get this to show up correctly the admin has to convert their "browser" time to UTC and enter an incorrect event time, in this case they'd have to enter the event time as April 3rd at 10pm, but set the timezone drop down to display MSK. This gets even more confusing with multiple admins using this across timezones, as they will have different views of the record.

This of course is ignoring that the browser timezone could be inaccurate for any variety of reasons.

### To Reproduce

n/a

### Payload Version

n/a

**Comment by paulpopus** [comment]
Moved this over to a discussion as it's a feature request! We can gather feed for this here

**Reply by ssyberg** [reply]
I thought you might say that 😆 

If it's not a bug, would you mind explaining what the payload way of doing this would be in it's current state?

**Reply by paulpopus** [reply]
There's a few ways to do it, the easiest I can think of right now would be to use a select list for the ability to select the timezone and then copy the core date field component and make a custom one and using the `useFormFields` [hook](https://payloadcms.com/docs/admin/hooks#useformfields) retrieve data from the timezone select.

Then when updating the field data for the date field you'll want to modify it to account for that selected timezone.

**Reply by ssyberg** [reply]
Yes, of course creating a custom component solves this, I meant what is the proper way to do this natively with the current implementation?

The only reason I'm pushing this is because it does feel like core Date field functionality is actually not functioning, this is why I submitted this as a bug. Not to get too philosophical but certainly some degree of common use case support is implicit in a feature being bug free, and the expected behavior definitely does not match what I'm seeing... as such calling this a feature request feels a little odd/dismissive to me.

**Reply by paulpopus** [reply]
It's a feature request because the date field is working as intended in this case and the ability to choose a timezone would be new functionality and this is one of the use cases of why we provide the ability for custom components.

I'm not trying to dismiss your suggestion as I'm in favour of it being added to core or a plugin.

**Reply by ssyberg** [reply]
Ok I'm sorry to be a pain in your side, but I find it difficult to let this go since this behavior is clearly incorrect, no one would say this is "how a date field works" 

- I set a date to April 3rd, 10pm
- It displays as April 4th, 8am

That is a bug.

**Reply by paulpopus** [reply]
I want to confirm the behaviour and your expectation here to make sure that there's no miscommunication between us, because if this is a bug then it obviously needs to be fixed.

You're setting the time as `April 3rd, 10pm` in the admin panel.
Then the API returns `April 4th, 8am`? Which is the UTC timestamp for the above when adjusted for your config. Is that right?

And this issue stems from having editors in different timezones who see different times when editing content?

If it's not what I described above then a reproduction would be helpful here to pinpoint the problem.
Not to sound condescending at all, we had a chat about this internally and I wanna clarify this thread.

**Reply by ssyberg** [reply]
Thanks for hearing me out!

I think if timezones are not supported, then they should not be affecting anything, I would simply remove the offset conversion in your `onChange` method and make it clear its all UTC (or like most other CMS' allow a config based setting for timezone). I don't actually think this is a result of multiple timezones for editors, I would suggest this is an actual bug for anyone editing that is not in UTC+0 because ultimately the the date/time that is entered is not the date/time that is stored.

**Reply by paulpopus** [reply]
I understand the need for timezone management in our date fields, at least to provide devs a bit more flexibility in how it's handled.

Until we can prioritise this for core I've added a date field with [timezone support here](https://github.com/NouanceLabs/payload-better-fields-plugin?tab=readme-ov-file#date-field) if it can help you in the interim? Feel free to provide feedback in that repo if you need something extra for it

**Reply by ssyberg** [reply]
Whoa cool, I didn't even know about that repo!

**Comment by iamlinkus** [comment]
Oh man, this date field timezone issue has been driving me nuts for the past two weeks. Having the admins of a website scattered throughout the world, it's impossible to handle this issue easily. The dates are set for events that are also international and are in different timezones, so this issue is on a whole another level (figure out the editor timezone, adjust while saving to negate the timezone difference and then also get a hook that adjusts the time when another editor is viewing the field in another timezone.

This is 100% a bug, because outside of the people who actually live in UTC+0, this field becomes completely unpredictable.

And a simple solution to this would simply be setting the default timezone that the date is saved in the db with. That's it.

**Reply by HappyEmu** [reply]
Yes, had a lot of issues already with our customers (admin panel users) living in different timezones, and things not happening when they should. This absolutely needs to be prioritized.

**Comment by paulpopus** [comment]
Thank you everyone for your feedback here! We'll be prioritising this feature and start working on it ASAP

**Reply by danielkoller** [reply]
Any update on this? This is really driving me nuts.

**Reply by r1tsuu** [reply]
@danielkoller PR is here for this feature - https://github.com/payloadcms/payload/pull/10896

**Reply by danielkoller** [reply]
Aweome, thanks!

**Reply by lars-berg-dxd** [reply]
+1 We've run into this issue with multiple clients now

**Reply by paulpopus** [reply]
This feature has been released a while ago https://payloadcms.com/docs/fields/date#timezones

**Reply by lars-berg-dxd** [reply]
hmmm.... maybe I'm misunderstanding how this works but I wasn't able to get it to work in 3.9.0:
<img width="762" alt="image" src="https://github.com/user-attachments/assets/85764022-bfaf-433e-91a2-08f9a7c92137" />

**Reply by paulpopus** [reply]
You should update to a newer version, I don't remember the exact version it went out but it was much later than `3.9.0`. latest is `3.39.1`

edit: ahh yeah it came in at `3.23.0`

**Reply by lars-berg-dxd** [reply]
thanks Paul, I'll give that a try. Sorry for resurfacing this issue that's been resolved

**Comment by paulpopus** [comment]
Hey all, this feature was released a while ago, my bad I forgot to update this discussion: https://payloadcms.com/docs/fields/date#timezones

Will lock it as resolved, if you encounter any bugs then please file an issue! Thank you everyone!

**Reply by samtigall** [reply]
even works with summer / winter time 👍 gj!

**Reply by paulpopus** [reply]
I should hope so! Our test suite tests from 3 distinct timezones and setting in other timezones with daylight and no daylight savings
![image](https://github.com/user-attachments/assets/6a4071d9-ca09-4304-bc3d-96b98360af61)
